### ISSUE TITLE
`MistralAttention`: where is the sliding window

### ISSUE THREAD

**Issue by fteufel** [huggingface/transformers#29777]
Hi,

I'm trying to understand the implementation of Mistral's attention in `MistralAttention`.
https://github.com/huggingface/transformers/blob/main/src/transformers/models/mistral/modeling_mistral.py#L195
It is my understanding that it should always be using local window attention. In `MistralFlashAttention2` this is very obvious, with `config.sliding_window` being used.

However, I'm not sure where the sliding window is used in the base `MistralAttention` without flash attention:

```python
class MistralAttention(nn.Module):
    """
    Multi-headed attention from 'Attention Is All You Need' paper. Modified to use sliding window attention: Longformer
    and "Generating Long Sequences with Sparse Transformers".
    """
```
but the forward pass simply reads
```python
attn_weights = torch.matmul(query_states, key_states.transpose(2, 3)) / math.sqrt(self.head_dim)
```
which I understand as full self attention.

Is the sliding window only used when running with Flash Attention, or am I missing something?
Thanks!

**Comment by amyeroberts** [comment]
cc @ArthurZucker @younesbelkada

**Comment by PenutChen** [comment]
I think the sliding window trick is based on masking?

**Comment by fteufel** [comment]
Thanks, I see. But wouldn't this throw away any computational efficiency gains expected from using a sliding window in the first place?

**Comment by PenutChen** [comment]
I have the same question. I think the sliding window has two aspects:

1. From the perspective of the attention mask, it essentially acts as a token-level sliding window that influences each token's view of the context.
2. From a kv-cache perspective, truncating the cache outside the window can improve computational efficiency.

Just my guess above.

**Comment by ArthurZucker** [comment]
Yes this would throw away the gains, and it is pretty much expected as the best way to use `sliding_window` is threw the `sdpa` or the `flash_attention` api, unless a rotating buffer is used. 

Closing as expected, feel free to discuss! 🤗

**Comment by fteufel** [comment]
Hi @ArthurZucker interesting - so `sdpa` actually exploits the local window structure of the attention mask in the backend?

**Comment by ArthurZucker** [comment]
It should if the mask is correctly passed yeah. New sdpa has the `sliding_window` argument anyway. Not sure it was correctly prepared before, important PR: #29407

**Comment by ehuaa** [comment]
> It should if the mask is correctly passed yeah. New sdpa has the `sliding_window` argument anyway. Not sure it was correctly prepared before, important PR: #29407

Did you mention this pr? https://github.com/pytorch/pytorch/pull/120143. torch 2.2 has no sliding_window argument in sdpa.

**Comment by ArthurZucker** [comment]
Sorry I meant the new SDPA codepath in transformers but it's not merged yet, yes as you say handles the mask

**Comment by MonolithFoundation** [comment]
Hi, then where does the attention mask needed to prepare from?

**Comment by MonolithFoundation** [comment]
I think it even might comes from dropout

**Comment by ArthurZucker** [comment]
Hey! here is where we do some of the work: https://github.com/huggingface/transformers/blob/093bebcdd9efea9626a0edf78f1cd8d256b1b652/src/transformers/models/mistral/modeling_mistral.py#L723-L730
